kernel: enable builds with BTF (BPF Type Format)  support

guidosarducci opened the pull request:
**Description:**

This PR enables support for BTF kernel debug information, and has been developed and tested over several months. The BTF information is generated from DWARF debug info and then embedded in the kernel for export via sysfs, where it can enhance kernel portability and introspection for BPF programs.

The commit series has three dependent parts:

1. _Clean up and add `ELF` related host libraries_
There are currently **three** active sources of the `libelf` host library on Linux, but most are unnecessary and cause conflicts/confusion; remove or disable these. We also need a `libdw` host library. Adding a host build of the `elfutils` package provides these required libraries.
See https://github.com/openwrt/packages/issues/9927#issuecomment-766299179 for prior discussion.
**Note**: `libelf` host library cleanup is now done: see bf4aa0c6a2f57951f995b2ac6f394c45ac852b9d and 29e2be69c4689f3438340f23a6eb5649516f0c6d.

2. _Addition of `dwarves` host tools_
A recent `pahole` (part of `dwarves`) is a Linux kernel build dependency when enabling BTF, and a host build is preferred due to inconsistent versions available across various Linux distros.

3. _Allow configuring BTF kernel and clean up related options_
Add option KERNEL_DEBUG_INFO_BTF and ensure continued working of existing KERNEL_DEBUG_INFO* options.

**Patches:**

Several patches were previously needed to support this PR, but all have since been fixed or submitted upstream and accepted.

**Testing:**

These changes have been build and run-tested across systems with 32-bit/64-bit word sizes and both big/little-endian byte orders, mainly using QEMU: malta/mips[32|64][eb|el], armvirt32, armvirt64, x86_64.

On-device testing uses the recently added `bpftool`: e.g.
```
root@OpenWrt:/# bpftool btf dump file /sys/kernel/btf/vmlinux format c
...
struct nf_conn {
        struct nf_conntrack ct_general;
        spinlock_t lock;
        u32 timeout;
        struct nf_conntrack_zone zone;
        struct nf_conntrack_tuple_hash tuplehash[2];
        long unsigned int status;
        u16 cpu;
        possible_net_t ct_net;
        struct hlist_node nat_bysource;
        struct {        } __nfct_init_offset;
        struct nf_conn *master;
        u_int32_t mark;
        struct nf_ct_ext *ext;
        union nf_conntrack_proto proto;
};
...
```

BTF info will add ~400KB to the size of a mips32 lzma-compressed debug kernel, similar to adding KALLSYMS info, noting that neither of these will be part of release kernel builds at present.

**Review/Distribution List:**

@feckert raised the prior point of [having a host build](https://github.com/openwrt/packages/issues/9927#issue-490168490) for `elfutils`, while @micmac1 and @neheb added discussion and confirmed the old BSD `tools/libelf` can be entirely removed on Linux (I confirmed as well).

@ldir-EDB0 has previously been pinged on the need for `tools\libelf` on Mac OSX (still looking for feedback). If the old BSD `libelf` can be replaced with `elfutils` on Mac OSX, some Makefile simplifications are possible.

@ffainelli, @nbd168 and @jow- have done past `tools\libelf` maintenance and so may have relevant history.

Pinging @dangowrt , @blogic and @PolynomialDivision as they've shown past interest in BPF support on OpenWRT.

Thanks for your feedback!

guidosarducci committed:
build: add option KERNEL_DEBUG_INFO_BTF

Generate BTF (BPF Type Format) information from DWARF debug info. This is
embedded in the kernel and exported via sysfs as /sys/kernel/btf/vmlinux.
BTF data enhances kernel portability and introspection for BPF programs.

Selecting this also enables the dwarves host package which provides the
pahole tool used for BTF encoding.

Test using: "bpftool btf dump file /sys/kernel/btf/vmlinux format c"

Also preserve existing behaviour for KERNEL_DEBUG_INFO_REDUCED by adding
it to the "Kernel build options" Kconfig.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
config: limit CONFIG_DEBUG_INFO to top-level generic configs

Remove redundant target-level entries, noting that these settings will be
configured from "Kernel build options" of Kconfig.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
tools/dwarves: add host package

dwarves is a set of tools that use the debugging information inserted in
ELF binaries by compilers such as GCC. Utilities in the dwarves suite
include pahole, which can be used to find alignment holes in structs and
classes, and also extracts other information such as CPU cacheline
alignment, helping pack those structures to achieve more cache hits.

These tools are also used to encode and read the BTF type information
format used with the bpf syscall, making this a Linux build dependency
when using kernel BTF information.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

PolynomialDivision commented:
> BTF info will add ~400KB to the size of a mips32 lzma-compressed debug kernel, similar to adding KALLSYMS info.

Maybe u want to use the SMALL_FLASH flag?

Thanks for adding this. I will have a closer look, soon! :)

adschm commented:
> Remove redundant target-level entries, noting that these settings will be
> configured from "Kernel build options" of Kconfig.

This is not "redundant". By using that term I assume that no effective change will be made. However, the option default to "no", and the options you remove are "yes".

So, please use a different description that will explain to you remove a default "yes" and hand this over to the user for configuration.

adschm commented on config/Config-kernel.in:
```diff
@@ -313,6 +313,21 @@ config KERNEL_DEBUG_INFO
+	bool "Enable additional BTF type information"
+	default n
+	depends on KERNEL_DEBUG_INFO
+	depends on !LINUX_4_14 && !LINUX_4_19
```
There is only 5.4 in master, so this line needs to be dropped.

guidosarducci replied:
Yes, when development started 5.4 was still in progress. Will remove now.

adschm commented on target/linux/layerscape/armv8_64b/config-5.4:
```diff
@@ -219,8 +219,6 @@ CONFIG_CRYPTO_RNG2=y
 CONFIG_DCACHE_WORD_ACCESS=y
 CONFIG_DEBUG_BUGVERBOSE=y
-CONFIG_DEBUG_INFO=y
-# CONFIG_DEBUG_INFO_REDUCED is not set
```
The CONFIG_DEBUG_INFO_REDUCED change should be moved to the subsequent patch, as this one should actually not be cosmetic here.

guidosarducci replied:
Yes, it's cleaner located in the next commit.

arkap commented:
Thank you for working on this. I've just today stumbled over the need for BTF info.
To test your PR I've built an image using your commits for an x86_64 platform. However, the BTF doesn't appear in `/sys/kernel/`. Am I missing something?

guidosarducci commented:
> Thank you for working on this. I've just today stumbled over the need for BTF info.
> To test your PR I've built an image using your commits for an x86_64 platform. However, the BTF doesn't appear in `/sys/kernel/`. Am I missing something?

I test periodically building on x86_64 (and just verified again), so that target should be fine.

Without seeing any of your verbose build logs, I would guess that your build environment is "dirty", and includes conflicting "leftovers" in `staging_dir` from previous builds of the old BSD-based `tools/libelf`. Doing a normal "make clean" doesn't clear out host tools or libs like this. Any conflicts can cause building `tools/dwarves` to fail, without which the kernel build will just skip creating BTF data.

Safest thing would be to do a "make dirclean", but you could also try "make tools/libelf/clean" on your tree *before* applying this PR.

guidosarducci commented:
> > BTF info will add ~400KB to the size of a mips32 lzma-compressed debug kernel, similar to adding KALLSYMS info.
> 
> Maybe u want to use the SMALL_FLASH flag?

That shouldn't be necessary since this setting is already off by default. A 400KB bump to kernel sizes would be quite large otherwise.

PolynomialDivision commented:
@guidosarducci  I can test at the weekend or maybe later today or tomorrow. :)

PolynomialDivision commented:
If someone else what to use curl to apply ;)
```
curl https://github.com/openwrt/openwrt/commit/f0fbeecd0a1f6ca941c02234903a392d915e0244.patch | git am && \
curl https://github.com/openwrt/openwrt/commit/d84c2e9d1e93c9fc0f265ebc18865c6920bac780.patch | git am && \
curl https://github.com/openwrt/openwrt/commit/0067c537d76fc946efd7d014892fc42bbca3497e.patch | git am && \
curl https://github.com/openwrt/openwrt/commit/e0babc24d7c57fd354d2b36f1b630346b8db6ce0.patch | git am && \
curl https://github.com/openwrt/openwrt/commit/8bb841ad5cc57b5569588c1a6510c9a076c00c94.patch | git am && \
curl https://github.com/openwrt/openwrt/commit/7908ea325772bb663af0daff5ba55c6853d1c1d6.patch | git am 

```

guidosarducci commented:
> @guidosarducci I can test at the weekend or maybe later today or tomorrow. :)

Thanks. Honestly, there's not too much to test, but having another confirmation on a non-virtual target is always good to have. At this stage, the show-stopping problems relevant to OpenWrt have been resolved/pushed to the upstream libbpf, dwarves and linux kernel repos.

And although non-Linux systems (e.g. BSDen) aren't the primary build platform, it would be nice to have some of the related feedback I requested, since ELF/DWARF libraries there are a bit of a mess from what I've read so far.

> If someone else want to use curl to apply ;)
> 
> ```
> curl https://github.com/openwrt/openwrt/commit/f0fbeecd0a1f6ca941c02234903a392d915e0244.patch | git am && \
> curl https://github.com/openwrt/openwrt/commit/d84c2e9d1e93c9fc0f265ebc18865c6920bac780.patch | git am && \
> curl https://github.com/openwrt/openwrt/commit/0067c537d76fc946efd7d014892fc42bbca3497e.patch | git am && \
> curl https://github.com/openwrt/openwrt/commit/e0babc24d7c57fd354d2b36f1b630346b8db6ce0.patch | git am && \
> curl https://github.com/openwrt/openwrt/commit/8bb841ad5cc57b5569588c1a6510c9a076c00c94.patch | git am && \
> curl https://github.com/openwrt/openwrt/commit/7908ea325772bb663af0daff5ba55c6853d1c1d6.patch | git am 
> ```

Hard to validate the above, and commits and ordering can change. I think most folks are better off using GitHub's automated patch series generation:

`curl -L https://github.com/openwrt/openwrt/pull/3855.patch |git am`

PolynomialDivision commented:
> Hard to validate the above, and commits and ordering can change. I think most folks are better off using GitHub's automated patch series generation:
> 
> `curl -L https://github.com/openwrt/openwrt/pull/3855.patch |git am`

Good to know that I can do this also. Much better... ;)

guidosarducci commented:
@ynezz Hi Petr, I forgot to mention you earlier as someone who's reviewed BPF-related work in the past. If you have a chance to look at this I'd appreciate it. This has been baking since last summer, and by now all the "big" problems I know of have been resolved upstream.

@arkap Hi Arne, did you track down your build issues? I'm curious if first doing `make dirclean` resolved things. Otherwise, looking at a verbose build log would be the next step.

PolynomialDivision commented:
@guidosarducci Can confirm everything is running on x86_64 PC Engines APU 2.

Thanks for your great work!

PolynomialDivision commented:
Also tested on ath79 WDR4300 (musl).

PolynomialDivision commented:
@guidosarducci Your first patch does not apply anymore correctly.

PolynomialDivision commented:
Suddenly, something is broken. Not even bpftools compile anymore.

arkap commented:
`bpf_helpers.h` mentions the cause:
```
/*
 * Note that bpf programs need to include either
 * vmlinux.h (auto-generated from BTF) or linux/types.h
 * in advance since bpf_helper_defs.h uses such types
 * as __u64.
 */
```
Looks like something is wrong with the way including `vmlinux.h` and BTF interact.

guidosarducci commented:
> @guidosarducci Your first patch does not apply anymore correctly.

Yes, there's been a lot of churn related to ABI version management, with a few lines rearranged in`package/libs/elfutils/Makefile`. I rebased the patch against latest master and verified everything still build and works fine, including `bpftools`.

> Suddenly, something is broken. Not even bpftools compile anymore.

@PolynomialDivision @arkap FYI, you can also check the latest snapshot builds to see that `bpftools` is indeed being built OK. Maybe a problem with your build/environment?

PolynomialDivision commented:
> @PolynomialDivision @arkap FYI, you can also check the latest snapshot builds to see that `bpftools` is indeed being built OK. Maybe a problem with your build/environment?

I will test again! :) Thanks.

arkap commented:
> @PolynomialDivision @arkap FYI, you can also check the latest snapshot builds to see that `bpftools` is indeed being built OK. Maybe a problem with your build/environment?

I've just cloned `master` and done nothing besides apply your PR and I'm still getting the same error.

guidosarducci commented:
> > @PolynomialDivision @arkap FYI, you can also check the latest snapshot builds to see that `bpftools` is indeed being built OK. Maybe a problem with your build/environment?
> 
> I've just cloned `master` and done nothing besides apply your PR and I'm still getting the same error.

I don't know what to tell you except that it sounds like an environment problem. It builds in master snapshot and I've rebuilt it a few times too without issue.

I think if you want to get to the root problem keep it simple: pick a master snapshot that has `bpftools`, check out the same commit and use their config to start a build. Confirm if you can reproduce the snapshot build, or if not try to understand why.

PolynomialDivision commented:
Compiled again with `Debian 10` and it works.

guidosarducci commented:
Pinging @ldir-EDB0 for review following up from IRC.

curtdept commented:
This is blocking 5.10 x64_64 builds :(

PolynomialDivision commented:
> This is blocking 5.10 x64_64 builds :(

Why?

curtdept commented:
> > This is blocking 5.10 x64_64 builds :(
> 
> Why?

libelf wont build a kernel > 5.8 with the old one we use its missing gelf_getsymshndx

guidosarducci commented:
> This is blocking 5.10 x64_64 builds :(

@curtdept  Right, you mean the lack of this PR is blocking... Threw me for a second there.

> libelf (the ancient version we use in tools now) wont build a kernel > 5.8 its missing gelf_getsymshndx. The elfutils version (this looks like its converting too) does however.
> https://github.com/torvalds/linux/blob/4e909124f8ed54b13e07e42273c3452b7deb5a0b/tools/objtool/elf.c#L378

Yes, the ancient version is only needed on BSDish systems and should be isolated, otherwise it can break software builds on Linux that need modern elf support, such as `dwarves`/`pahole`.

Unfortunately, as you can see in the OP I've pinged several relevant reviewers with little effect, and tried on IRC too. Great if you could help nudge anyone to look as well. Is there an x86_64 5.10 PR to mention this?

curtdept commented:
> > This is blocking 5.10 x64_64 builds :(
> 
> @curtdept Right, you mean the lack of this PR is blocking... Threw me for a second there.
> 
> > libelf (the ancient version we use in tools now) wont build a kernel > 5.8 its missing gelf_getsymshndx. The elfutils version (this looks like its converting too) does however.
> > https://github.com/torvalds/linux/blob/4e909124f8ed54b13e07e42273c3452b7deb5a0b/tools/objtool/elf.c#L378
> 
> Yes, the ancient version is only needed on BSDish systems and should be isolated, otherwise it can break software builds on Linux that need modern elf support, such as `dwarves`/`pahole`.
> 
> Unfortunately, as you can see in the OP I've pinged several relevant reviewers with little effect, and tried on IRC too. Great if you could help nudge anyone to look as well. Is there an x86_64 5.10 PR to mention this?

There is a WIP one that is incomplete and looks abandoned, I'll start one I was working on.

PolynomialDivision commented:
> > > This is blocking 5.10 x64_64 builds :(
> > 
> > 
> > @curtdept Right, you mean the lack of this PR is blocking... Threw me for a second there.
> > > libelf (the ancient version we use in tools now) wont build a kernel > 5.8 its missing gelf_getsymshndx. The elfutils version (this looks like its converting too) does however.
> > > https://github.com/torvalds/linux/blob/4e909124f8ed54b13e07e42273c3452b7deb5a0b/tools/objtool/elf.c#L378
> > 
> > 
> > Yes, the ancient version is only needed on BSDish systems and should be isolated, otherwise it can break software builds on Linux that need modern elf support, such as `dwarves`/`pahole`.
> > Unfortunately, as you can see in the OP I've pinged several relevant reviewers with little effect, and tried on IRC too. Great if you could help nudge anyone to look as well. Is there an x86_64 5.10 PR to mention this?
> 
> There is a WIP one that is incomplete and looks abandoned, I'll start one I was working on.

I thiank I'm the PR. I had no time anymore to go on...

curtdept commented:
> > > > This is blocking 5.10 x64_64 builds :(
> > > 
> > > 
> > > @curtdept Right, you mean the lack of this PR is blocking... Threw me for a second there.
> > > > libelf (the ancient version we use in tools now) wont build a kernel > 5.8 its missing gelf_getsymshndx. The elfutils version (this looks like its converting too) does however.
> > > > https://github.com/torvalds/linux/blob/4e909124f8ed54b13e07e42273c3452b7deb5a0b/tools/objtool/elf.c#L378
> > > 
> > > 
> > > Yes, the ancient version is only needed on BSDish systems and should be isolated, otherwise it can break software builds on Linux that need modern elf support, such as `dwarves`/`pahole`.
> > > Unfortunately, as you can see in the OP I've pinged several relevant reviewers with little effect, and tried on IRC too. Great if you could help nudge anyone to look as well. Is there an x86_64 5.10 PR to mention this?
> > 
> > 
> > There is a WIP one that is incomplete and looks abandoned, I'll start one I was working on.
> 
> I think I'm the PR. I had no time anymore to go on...

Ha no worries, wasn't meant to be shade :) just happened to have the day off myself, thought I would play haha

guidosarducci commented:
Re-pinging some reviewers requested in the OP: @ldir-EDB0 @nbd168 @dangowrt @jow.

TDLR: These BPF related improvements have been marinating/testing since last summer, and the prerequisite elf library fixes have broader impact, even addressing problems seen updating x86_64 to 5.10. Please have a look if possible.

curtdept commented:
Requesting this PR be broken up into elfutils changes and dwarves changes as the can be discrete components.

guidosarducci commented:
The changes are already discrete and logical, separated into commits, and it's straightforward to pull some or all them into master. The current obstacle is getting reviewed by relevant committers, rather than any shortage of PRs, so pinging again some of the devs I've highlighted above would best help in getting attention, as I mentioned in https://github.com/openwrt/openwrt/pull/3942#issuecomment-787110048. Would you help raise this need where you can? Thanks!

curtdept commented:
> The changes are already discrete and logical, separated into commits, and it's straightforward to pull some or all them into master. The current obstacle is getting reviewed by relevant committers, rather than any shortage of PRs, so pinging again some of the devs I've highlighted above would best help in getting attention, as I mentioned in [#3942 (comment)](https://github.com/openwrt/openwrt/pull/3942#issuecomment-787110048). Would you help raise this need where you can? Thanks!

Don't really have any sway, just PR occasionally but feedback wise trying this. The elfutils changeover resulted in a non-bootable image for me. Haven't spent pretty much any time with it, but I am guessing there are some kind of cross platform compiling quirks with grub and that change on x86.

guidosarducci commented:
> trying the elfutils changeover resulted in a non-bootable image for me. Haven't spent pretty much any time with it, but I am guessing there are some kind of cross platform compiling quirks with grub and that change on x86.

Note that you need to clean out the the old `tools/libelf` artifacts since they can interfere with subsequent builds: so perform a `make dirclean`, or a `make tools/libelf/clean` ***before*** applying these commits, or simply clean up manually. I'll try to confirm building x86_64 myself if that helps.

I also connected with @pprindeville and asked if he could look at the x86_64 build and my libelf fixes.

curtdept commented:
> > trying the elfutils changeover resulted in a non-bootable image for me. Haven't spent pretty much any time with it, but I am guessing there are some kind of cross platform compiling quirks with grub and that change on x86.
> 
> Note that you need to clean out the the old `tools/libelf` artifacts since they can interfere with subsequent builds: so perform a `make dirclean`, or a `make tools/libelf/clean` _**before**_ applying these commits, or simply clean up manually. I'll try to confirm building x86_64 myself if that helps.
> 
> I also connected with @pprindeville and asked if he could look at the x86_64 build and my libelf fixes.

Ya builds good, won't boot though, fresh repo clone.

guidosarducci commented:
@httpstorm Since you're a macOS user, I'd appreciate if you could have a look at the first 3 commits of this PR, and also the discussion in https://github.com/openwrt/packages/issues/9927#issuecomment-766299179 regarding `libelf` and macOS.

httpstorm commented:
@guidosarducci I'll have a look later today.

httpstorm commented:
@guidosarducci 
The `master-add-btf-kernel` branch builds correctly under macOS 11, starting from a blank directory, and adding the `.config` used by my own builds. That is without selecting the option mentioned below… Probably not worth mentioning, I added some of my patches that fix unrelated build errors in the macOS build environment. All but one are already accepted and part of the master. The produced image runs on WRT3200. I am not familiar with the technical part being discussed here, but if you make a list of steps you want me to perform and test, I might be able to help. The size of the generated image is the same as the size of my own builds. I would assume BTF is not enabled be default, is it?

Starting again from a fresh directory, `make -j 16` failed after I selected:
- Global build settings, Kernel build options, Enable additional BTF type information

```
make V=sc
…
make[3]: Entering directory '/Volumes/btf/openwrt/tools/mpc'
make[3]: Leaving directory '/Volumes/btf/openwrt/tools/mpc'
time: tools/mpc/compile#0.06#0.06#0.15
make[2]: *** No rule to make target 'tools/elfutils/compile', needed by 'tools/dwarves/compile'.  Stop.
make[2]: Leaving directory '/Volumes/btf/openwrt'
make[1]: *** [tools/Makefile:166: /Volumes/btf/openwrt/staging_dir/host/stamp/.tools_compile_yyynyynnyyyyynyyyyyynyynnyyyynyyyyyyyyyyyyyyyynynnyyyyyyy] Error 2
```

I suppose more work is required on macOS? Good luck!

guidosarducci commented:
@httpstorm 
> The `master-add-btf-kernel` branch builds correctly under macOS 11, starting from a blank directory, and adding the `.config` used by my own builds. That is without selecting the option mentioned below…

Excellent! Thanks for confirming, this is exactly what I wanted to know: that the PR doesn't break the default build on macOS. Unfortunately, macOS builds rely on `tools/libelf`, which shouldn't be used on Linux systems and is thus hidden by commit 2. I wanted to be sure, however, that hiding it from Linux systems didn't affect macOS builds.

> I am not familiar with the technical part being discussed here, but if you make a list of steps you want me to perform and test, I might be able to help. The size of the generated image is the same as the size of my own builds. I would assume BTF is not enabled be default, is it?

Correct, the extra BTF debug information is not created by default, otherwise the kernel size would increase slightly. Also, I've tried to make the commit descriptions clear and detailed, in case you're looking for more info (let me know if they're not!). 

> Starting again from a fresh directory, `make -j 16` failed after I selected:
> 
>     * Global build settings, Kernel build options, Enable additional BTF type information
> 
> 
> ```
> make V=sc
> …
> make[3]: Entering directory '/Volumes/btf/openwrt/tools/mpc'
> make[3]: Leaving directory '/Volumes/btf/openwrt/tools/mpc'
> time: tools/mpc/compile#0.06#0.06#0.15
> make[2]: *** No rule to make target 'tools/elfutils/compile', needed by 'tools/dwarves/compile'.  Stop.
> make[2]: Leaving directory '/Volumes/btf/openwrt'
> make[1]: *** [tools/Makefile:166: /Volumes/btf/openwrt/staging_dir/host/stamp/.tools_compile_yyynyynnyyyyynyyyyyynyynnyyyynyyyyyyyyyyyyyyyynynnyyyyyyy] Error 2
> ```

Sadly, this is entirely expected since macOS uses the old `tools/libelf` dating from 2009. Kernel features like BTF require host tools (e.g. `dwarves`) which can't be built with `tools/libelf`, and trying to compile newer kernels will also fail with this `tools/libelf` (see https://github.com/openwrt/openwrt/pull/3855#issuecomment-786903213).

> I suppose more work is required on macOS? Good luck!

I think the future of macOS as a build platform will depend on replacing `tools/libelf` with the more modern `elfutils` or equivalent. Everything I've researched so far says this is difficult; I've found no `elfutils`-based software packages for macOS anywhere. In light of all that, I think we must treat `BTF` as "not-supported" on macOS for the time being.

Someone with access to and knowledge of the platform would be needed to take things further. I've shared some of my findings with @ldir-EDB0 but haven't had any feedback. Maybe you two can talk?

Thanks again for your help.

httpstorm commented:
@guidosarducci 
I'm glad if I helped! The commit descriptions look good, though the topic is beyond my field of experience.
@ldir-EDB0 Kevin?

guidosarducci commented:
@httpstorm Hi Georgi, there was another macOS test I could use your help with. I'd like to determine if `tools/libelf` is still needed to build on macOS. When you have a chance, could you try:
1. Running `make tools/libelf/clean` to remove existing `libelf` artifacts from `staging_dir`. Please double-check all the `libelf` pieces are removed.
2. Apply the following patch or edit to disable `tools/libelf` on macOS:
```
diff --git a/tools/Makefile b/tools/Makefile
index 67e5de2b50..2ebbad42f7 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -82,7 +82,7 @@ endif

 ifneq ($(HOST_OS),Linux)
   $(curdir)/libelf/compile := $(curdir)/libtool/compile
-  tools-$(BUILD_TOOLCHAIN) += libelf
+  tools-n += libelf
 else
   $(curdir)/elfutils/compile := $(curdir)/m4/compile
   $(curdir)/elfutils/compile := $(curdir)/zlib/compile
```
3. Run a full rebuild without `tools/libelf`: `make clean` then `make V=s` and please save the logs to pastebin, etc. to confirm.

Thanks again for your help!

guidosarducci commented:
Yes, please try that on master. Sorry, my patch above was against this branch but I think you understand how to edit `tools\Makefile` manually, or use this patch:
```
diff --git a/tools/Makefile b/tools/Makefile
index 1d7da0756d..2d6eb43cc9 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -28,7 +28,7 @@ tools-y += mklibs mm-macros mtd-utils mtools padjffs2 patch-image
 tools-y += patchelf pkgconf quilt squashfskit4 sstrip xxd zip zlib zstd
 tools-$(BUILD_B43_TOOLS) += b43-tools
 tools-$(BUILD_ISL) += isl
-tools-$(BUILD_TOOLCHAIN) += expat gmp libelf mpc mpfr
+tools-$(BUILD_TOOLCHAIN) += expat gmp mpc mpfr
 tools-$(CONFIG_TARGET_apm821xx)$(CONFIG_TARGET_gemini) += genext2fs
 tools-$(CONFIG_TARGET_ath79) += lzma-old squashfs
 tools-$(CONFIG_TARGET_mxs) += elftosb sdimage
```

We just want to make sure `tools/libelf` is cleaned before disabling and building.

curtdept commented:
Well good news is I can clear this from my no have traced that back to an unrelated grub issue with higher GCC optimization levels being inherited from target_cflags

httpstorm commented:
@guidosarducci Hi Tony!

Everything compiles correctly with that patch applied. I also did `rm -rf tools/libelf` during the second build.
Seems like the build logs are "a bit" over the limits of pastebin.
Fortunately we can use my web server project for hosting `httpstorm`:
http://httpstorm.com/tmp/guidosarducci/

The short log is `master` with your patch applied, and `make -j 16`.
The log is `master` with your patch applied, `rm -rf tools/libelf` and `make V=sc -j 16`.
Both were performed using a freshly downloaded master, and the `.config` used for my own router, to make things more realistic. I also make sure my own software builds correctly.

Good luck!

guidosarducci commented:
Thank you, Georgi. That sounds really promising!

I'm a little worried that the previous `libelf` pieces may not have been removed before your build. Could you please confirm you did `make tools/libelf/clean` *before* applying the patch above and rebuilding? That was step no. 1 from https://github.com/openwrt/openwrt/pull/3855#issuecomment-796062532 above.

Doing `rm -rf tools/libelf` will only remove the `tools/libelf` makefile and patches, but would leave `libelf` library and header files under `./staging_dir/host/` that the next build could find/use.

BTW, the safest thing would be to do a `make dirclean` before your build, but that would delete your toolchain, etc. in additon to `./staging_dir/` and I thought you'd want to avoid that.

I'll also ask @ldir-EDB0 if he can confirm building without the `tools/libelf` libs is successful. Having extra verification would be good before remove that entirely.

httpstorm commented:
@guidosarducci 
Hello Tony! I did specifically mention building freshly cloned sources in a new and blank virtual disk. I always do this before making pull-requests, to avoid the risk of any interference from left-overs. That's even better than dirclean, and I did it twice. I cannot say if any other packages require libelf. I can only confirm that all packages I use in my router compile correctly without it on the current master.

guidosarducci commented:
Ah, sorry I overlooked the "fresh clone" comment and thanks for clarifying. This is great, and means we can drop tools/libelf unconditionally. I'll pull that out and make a separate PR doing that.

httpstorm commented:
@guidosarducci 
`libelf` is available from brew on macOS, and ready installed on my computer.
`elfutils` is not available. I remember I had build errors related to it during the previous test you asked me to do https://github.com/openwrt/openwrt/pull/3855#issuecomment-789915300
```
brew list |grep libelf
libelf

brew search elfutils
Error: No formulae or casks found for "elfutils".
```

guidosarducci commented:
@httpstorm Interesting! Out of curiosity, how are you using brew's `libelf`? Do your OpenWrt builds see/find that `libelf`? I'm wondering if, after removing the old `tools/libelf`, the "brew" libelf is being picked up in the fresh clones.

Also, if a `libelf` is needed in the future, do you think the brew package could become a system prerequisite on macOS? i.e. is it something stable and well maintained?

httpstorm commented:
@guidosarducci 
I think I installed `libelf` as a prerequisite to build OpenWRT in the past, as there were related errors. I am not familiar with its purpose. And it is not used by any of my projects.

guidosarducci commented:
> @guidosarducci
> I think I installed `libelf` as a prerequisite to build OpenWRT in the past, as there were related errors. I am not familiar with its purpose. And it is not used by any of my projects.

I wasn't aware there was another `libelf` installed, so I'm now rather confused about what is needed to compile on MacOS. My understanding was that after c3e31b6a9b and  5f8e587240 no host `libelf` should be needed to build the kernel/tools.

Were the related errors you used to see before these commits were added? Would it be possible to try a fresh build where both `tools/libelf` and the brew package `libelf` are removed? Hopefully that would give us a clear answer... 🤞

httpstorm commented:
@guidosarducci 
Hi Tony!
Honestly I don't remember how long ago it was. Probably before the commits you mentioned, because after that everything should compile without `libelf`. To clear any doubt on that matter, I uninstalled `libelf`, then starting from a blank directory, I cloned master, applied the patch which disables `libelf`, deleted `tools/libelf`, and then compiled OpenWRT successfully. That should prove the brew version I previously had installed as obsolete.

List of currently installed packages from brew and MacPorts:
https://pastebin.com/TC7QbNke

List of references to libelf in the current OpenWRT master:
https://pastebin.com/nYiQDxFw

guidosarducci commented:
> Hi Tony!
> Honestly I don't remember how long ago it was. Probably before the commits you mentioned, because after that everything should compile without `libelf`. To clear any doubt on that matter, I uninstalled `libelf`, then starting from a blank directory, I cloned master, applied the patch which disables `libelf`, deleted `tools/libelf`, and then compiled OpenWRT successfully. That should prove the brew version I previously had installed as obsolete.
> 
> List of currently installed packages from brew and MacPorts:
> https://pastebin.com/TC7QbNke
> 
> List of references to libelf in the current OpenWRT master:
> https://pastebin.com/nYiQDxFw

Hey Georgi, that's just perfect and demonstrates we can safely remove `tools/libelf` without causing MacOS users any problems (my main worry). Thanks for your patience and help sorting this out!

@httpstorm @neheb I'll PR that change separately. Is it OK if I add "Tested-by:" lines for MacOS and Linux referring to you both?

neheb commented:
Sure.

No comment on tools/libelf

httpstorm commented:
@guidosarducci 
I am glad to help you Tony, and thankful to see responsible people who test carefully to avoid issues for the others. 😊

guidosarducci commented:
Rebased against latest master and dropped two commits which have now been accepted: bf4aa0c6a2 and 29e2be69c4.

aparcar commented on tools/elfutils/Makefile:
```diff
@@ -0,0 +1,42 @@
+#
```
Please use SPDX conform header.

guidosarducci replied:
Right, will change both this and for `tools/dwarves/Makefile`. 

BTW, is SPDX a recent policy thing? I noticed most  utilities in `tools/` aren't using it and followed their example.

aparcar replied:
It's a recent thing and I'll just try to enforce it wherever something new is added, it's not a official policy.

aparcar commented on target/linux/generic/config-5.4:
```diff
@@ -1161,7 +1161,7 @@ CONFIG_DEBUG_FS=y
 # CONFIG_DEBUG_INFO_BTF is not set
 # CONFIG_DEBUG_INFO_DWARF4 is not set
-CONFIG_DEBUG_INFO_REDUCED=y
+# CONFIG_DEBUG_INFO_REDUCED is not set
```
Shouldn't this be added to `config-5.10` too?

guidosarducci replied:
Yes, thanks for noting. The patch set is 8 months old and I see things have changed underneath it.

Will update that, and also some archs that have since added `CONFIG_DEBUG_INFO=y`.

aparcar replied:
Thank you.

guidosarducci commented on tools/elfutils/Makefile:
```diff
@@ -0,0 +1,42 @@
+	--disable-debuginfod \
+	--without-bzlib
+
+#define Host/Uninstall
+#	-$(MAKE) -C $(HOST_BUILD_DIR) uninstall
+#endef
```
I'd been looking at trade-offs in different uninstall processes a while ago. I'll restore that section as it's simpler. Thanks.

guidosarducci commented:
@aparcar Thanks, I've updated based on your comments. Do you have a view on just adding a new `libdw-dev` build system prerequisite rather than including an `elfutils` host build in this PR? It would simplify the PR, but I don't know the related effort e.g. installing a new dev lib on the buildbots.

@httpstorm Hi Georgi, I've added back the guard to not build `elfutils` on MacOS after I mistakenly dropped it while removing the `tools\libelf` code.

aparcar commented:
How long does the `elfutils` build take? I'd rather have it in tree if it changes a lot between distros and distro versions.

guidosarducci commented:
> How long does the `elfutils` build take? I'd rather have it in tree if it changes a lot between distros and distro versions.

From what I've seen, `elfutils` is not a fast moving target and distros have had time to package recent stable versions. I've built BTF-aware kernels using `libelf-dev` and `libdw-dev` from both Ubuntu LTS 18 and 20, so I think the distro route would be fine.

The `dwarves` package absolutely needs a host build however. The needed BTF and cross-compilation support is very recent, and distro versions can be all over the map.

Build time on my PC (Intel(R) Core(TM) i7 CPU 860 @ 2.80GHz):
```
kodidev:~/openwrt-project$ make tools/elfutils/clean && time make -j8 tools/elfutils/compile
...
 make[1] tools/elfutils/compile
 make[2] -C tools/flock compile
 make[2] -C tools/xz compile
 make[2] -C tools/sed compile
 make[2] -C tools/tar compile
 make[2] -C tools/patch compile
 make[2] -C tools/pkgconf compile
 make[2] -C tools/libressl compile
 make[2] -C tools/cmake compile
 make[2] -C tools/zstd compile
 make[2] -C tools/ccache compile
 make[2] -C tools/m4 compile
 make[2] -C tools/zlib compile
 make[2] -C tools/lzma compile
 make[2] -C tools/elfutils compile

real    0m34.683s
user    0m35.958s
sys     0m10.064s
```

aparcar commented:
I wonder if we should a add a build requirement system that only kicks in with certain settings, similar to `qemu-utils` which is only required for some x86 special VDI stuff. If `elfutils` is only required for `dwarves`, why not tangle a nice error message around it letting the user know to install it if `dwarves` is desired? This way we don't add more *hard* host dependencies but only if someone actually needs the feature.

guidosarducci commented:
To be clear, `elfutils` isn't monolithic. Most distros build from that source and package 3 libs: `libelf-dev`, `libdw-dev` and `libasm-dev`.

BTF support (i.e `dwarves`) requires both `libelf-dev` and `libdw-dev`, but OpenWrt *already* has a dependency on `libelf-dev` (for building kernels with STACK_VALIDATION), so adding a `libdw-dev` doesn't seem too onerous. My thought with building host `elfutils` was to "kill two birds", while also allowing us to drop the `libelf-dev` system prerequisite *and* avoid updating buildbots with `libdw-dev`.

I'm not familiar with the `qemu-utils` history, but while it may be x86-specific, BTF support is across all archs.

One other thought is that since DEBUG_INFO_BTF is disabled by default and not enabled with ALL_KMODS, buildbots won't build it anyway. Perhaps we could just add a build-time prerequisite check for `libdw-dev` only if DEBUG_INFO_BTF is set? 

From previous discussion with @ynezz, it seems `libelf-dev` also lacks a prerequisite check, so maybe add one there unconditionally? 

End result could be that we add a `libdw-dev` system prerequisite, build-time checks for both `libelf-dev` and `libdw-dev`, and don't need to update buildbots. What do you guys think? Would that work?

aparcar commented:
Works for me

httpstorm commented:
@guidosarducci
Hello Tony, I am sorry for the late replay! I've been quite busy lately.
Here are the results of building your branch `master-add-btf-kernel` on macOS 11:
- Global build settings, Kernel build options, Enable additional BTF type information: disabled - builds correctly
- Global build settings, Kernel build options, Enable additional BTF type information: enabled - fails:
```
make -j 2 V=sc
…
make[2]: *** No rule to make target 'tools/elfutils/compile', needed by 'tools/dwarves/compile'.  Stop.
```

guidosarducci committed:
build/prereq: add libelf and libdw requirements

Add build prerequisite checks for ELF and DWARF development headers and
libraries. Host libelf is already an OpenWrt requirement but unchecked,
while libdw is needed by tools/dwarves and used to build kernels with BTF
debug information.

Include a guard to reflect that libdw is only used building tools/dwarves.
Since BTF debug kernels are disabled by default, no immediate buildbot
changes are needed.

Suggested-by: Petr Štetiar <ynezz@true.cz>
Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci commented:
@httpstorm I had only pinged you as an FYI that I temporarily broke something on MacOS, not to pressure you for more testing! Any apologies are mine, and thanks for re-testing that.

httpstorm commented:
No worries. Good luck! 😊

guidosarducci commented:
@aparcar @ynezz Following up on my proposal above and Petr's comment https://github.com/openwrt/openwrt/pull/3989#issuecomment-799186878, I've added build prerequisite checks for both `libelf` and `libdw`, dropped the host build of `elfutils`, and done testing with/without the deps installed.

Following merge, I would update the OpenWrt wiki to add a `libdw-dev` requirement next to the existing `libelf-dev` one. However, with this update no immediate change to buildbots is needed since BTF debug kernels are disabled by default.

Please have a look and let me know what you think. Hopefully we can support this kernel 5.4 feature soon.

aparcar commented on include/prereq-build.mk:
```diff
@@ -53,6 +53,18 @@ $(eval $(call TestHostCommand,ncurses, \
 	echo 'int main(int argc, char **argv) { initscr(); return 0; }' | \
 		gcc -include ncurses.h -x c -o $(TMP_DIR)/a.out - -lncurses))
+
+$(eval $(call TestHostCommand,libelf-dev, \
```
I've been building OpenWrt for some time now and never needed libelf.{so,h}, what am I missing?

guidosarducci replied:
It's been a prerequisite for a long time ((https://openwrt.org/docs/guide-developer/build-system/install-buildsystem), but without an explicit check I suspect it's missing for some people. Previously, it supports building kernels with STACK_VALIDATION enabled (for at least x86/64 target), but without it the kernel build doesn't fail, just prints a warning that most folks don't notice:
```
warning: Cannot use CONFIG_STACK_VALIDATION=y, please install libelf-dev, libelf-devel or elfutils-libelf-devel
```
I added this check unconditionally because we now will have two users for `libelf-dev`.

aparcar replied:
I just checked on the latest buildbot container image and it's missing there as well.

```bash
buildbot@c2789e94c2f6:/$ apt list --installed | grep libelf
libelf1/stable,now 0.176-1.1 amd64 [installed,automatic]
```

The `libelf-dev` package is not installed by default

Try locally via:
```
user@dawn:~/src/openwrt/21.02$ sudo docker run -it --rm registry.gitlab.com/openwrt/buildbot/buildworker apt list --installed | grep libelf
libelf1/stable,now 0.176-1.1 amd64 [installed,automatic]
```

aparcar replied:
Please create a PR here to add that package: https://gitlab.com/openwrt/buildbot

Ping @ynezz

guidosarducci replied:
First gitlab PR, no idea if this is right: https://gitlab.com/openwrt/buildbot/-/merge_requests/1

ynezz replied:
It's not correct, nobody is going to review patches on 3 different platforms. GitLab is currently only used for evaluation purposes. Ideally this should go over mailing list, thanks.

ynezz replied:
> Please create a PR here to add that package: https://gitlab.com/openwrt/buildbot

I think, that openwrt.org/submitting-patches is good enough.

aparcar replied:
Sorry for the extra loop. With added CI to buildbot i figured it's best to create it there directly

ynezz replied:
Mirroring is setup from git.openwrt.org to gitlab.com

guidosarducci replied:
> It's not correct, nobody is going to review patches on 3 different platforms. GitLab is currently only used for evaluation purposes. Ideally this should go over mailing list, thanks.

I can assure you nobody wants to submit patches on 3 different platforms either, as I've had to do in just the last week.

><aparcar[m]> guidosarducci: you can use your github account to login, if it's to much of a fuzz I'll do it

@aparcar Since you offered on IRC to take care of this, I'll ask if you can do that now. Thanks, it's been too much of a fuzz.

guidosarducci replied:
@aparcar Thank you for updating the buildbots. I'll close this off assuming there are no other concerns.

aparcar commented:
@ynezz please review this. Your concern was partly the lack of Mac OS X support, which seems working as tested by @httpstorm. @jow- suggested to have it as host tool, I'd be fine if we go that path even though it means more maintenance on our side.

stintel reviewed (approved):
From a quick glance, without reading the entire history of conversations or without runtime testing, I welcome changes like this and as we've already branched 21.02, I see no reason to hold this back.

guidosarducci commented:
@adschm I'd appreciate if you could add the "kernel" tag for this to be clearer. Thanks.

@ynezz I understand you still had some reservations about this PR. Could you elaborate so we can discuss?

guidosarducci commented:
Updates `tools/dwarves` to upstream v1.21 and rebased branch.

Retested BTF-enabled image generation and accessing using `bpftool` on multiple archs:

- build-tested on `mpc85xx/p1010/sophos_red-15w-rev1`
- build and run-tested using QEMU on `mips32/be`, `mips64el`, `armvirt/32`, `armvirt/64`, `x86_64`

bobafetthotmail commented:
Just checking, I didn't follow this PR but I've noticed in my snapshot builds (not from this PR) that I get a warning line about missing libraries for CONFIG_STACK_VALIDATION
`warning: Cannot use CONFIG_STACK_VALIDATION=y, please install libelf-dev, libelf-devel or elfutils-libelf-devel`

With some searching on github I found that you were also doing something on this.

it seems you are already dealing with this problem in this PR with the commit 
`build/prereq: add libelf and libdw requirements `
so do you confirm that it will fix that warning and I don't need to open a PR specifically about that?

guidosarducci commented:
> Just checking, I didn't follow this PR but I've noticed in my snapshot builds (not from this PR) that I get a warning line about missing libraries for CONFIG_STACK_VALIDATION
> `warning: Cannot use CONFIG_STACK_VALIDATION=y, please install libelf-dev, libelf-devel or elfutils-libelf-devel`
> 
> With some searching on github I found that you were also doing something on this.
> 
> it seems you are already dealing with this problem in this PR with the commit
> `build/prereq: add libelf and libdw requirements `
> so do you confirm that it will fix that warning and I don't need to open a PR specifically about that?

The warning is actually an old one that many people get but don't notice, having not installed `libelf-dev` although it's been a listed requirement for some time (e.g. on the wiki). The `build/prereq` commit in this PR just enforces the requirement so that people won't end up missing that warning.

guidosarducci commented:
> 
> @ynezz I understand you still had some reservations about this PR. Could you elaborate so we can discuss?

Ping @ynezz ?

guidosarducci commented:
@aparcar Hi Paul, is there anything remaining to do for this PR before merging? I'm assuming "no" since I've received no responses  on IRC or within this PR to questions of any outstanding concerns. Thanks!

aparcar commented:
@guidosarducci I think there where some concerns with adding both libs as a build requirement but rather have them in tools, these came from @jow- and/or @ynezz and should be addressed.

I'm fine with adding this if it's really valuable, but then again I have literally no clue what this does so someone who touched some similar (@ldir-EDB0 ?) should proceed merging this.

aparcar commented:
@dangowrt ping this is your playground

PolynomialDivision commented:
Can I somehow help testing? Finally, I have time again to look at ebpf. :D

dangowrt commented:
I will give it a shot now, anyway working on GCC-10 as BPF compiler (so we won't need LLVM/Clang).

PolynomialDivision commented:
> I will give it a shot now, anyway working on GCC-10 as BPF compiler (so we won't need LLVM/Clang).

Great news! :)

dangowrt commented:
FYI, see https://git.openwrt.org/?p=openwrt/staging/dangole.git;h=refs/heads/gcc10-bpf

What's working already:
 * bpf-unknown-none-* gets build and looks like it'd be working

What's not working (yet):
 * For some reason `--with-headers=` doesn't have the desired effect.
 * something still breaks if building with regular toolchain once gcc-bpf is present (possibly file/header collision, it got very late and I wasn't able to figure it out yet, review with that in mind is very welcome)

Minor issues:
 * -fhonor-copts will have to be taken care of, ie. set global CFLAGS_BPF=-fhonor-copts somewhere (where?)

dangowrt commented:
I think the biggest part of the work needed to have native BPF support with the OpenWrt toolchain is done. I got additional builds for gcc/bpf and binutils/bpf integrated in the toolchain build and that starts to be good enough to build at least parts of bpf-examples with plausible (but untested) outcomes. I'll continue in the next day to do the remaining polishing (headers, CFLAGS, ...) and then we should be good to go :)

That being said, I didn't have a chance yet to actually test BTF format stuff presented in this patch series (apart from pulling it into my staging tree and seeing it build well).

See my updated staging tree for the details:
https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=shortlog;h=refs/heads/gcc10-bpf

stintel commented:
Closing this in favour of #9362

ba1ajinaidu-ushur commented:
@stintel was this merged? I don't see BTF config options in `config/Config-kernel.in`. On v22.03.3

stintel commented:
I don't think this was backported to the 22.03 branch.

ba1ajinaidu-ushur commented:
Is this being maintained in any active branch?

stintel commented:
This was committed to master in May 2022 (531e0399cd7fca9e1a88520b817ac036da204c1c).

stintel commented:
> seems like it not enable btf by default. but just add a option about it. is there any plan to enable it?

This requires KERNEL_DEBUG_INFO_REDUCED=y which will cause bigger images, so this is unlikely to be enabled by default.

ynezz commented on include/prereq-build.mk:
```diff
@@ -53,6 +53,18 @@ $(eval $(call TestHostCommand,ncurses, \
+
+$(eval $(call TestHostCommand,libelf-dev, \
+	Please install ELF dev libs and headers. (Missing libelf.so or libelf.h), \
+	echo 'int main(int argc, char **argv) { return 0; }' | \
```
@guidosarducci I'm getting following reports occasionally:

```
fatal error: gelf.h: No such file or directory
```

and the solution is always to install `libelf-dev`, do you've any idea why this part was not merged? Was it causing some issues for example? I'm simply toying with an idea to merge it.
